### PR TITLE
[Ready to Review] Implement Multi-Interface Initialisation and Management in CLI

### LINKED ISSUE #83: Add Docstrings to `HardwareInterface` class
Add docstrings to `HardwareInterface` abstract base class in `core/hardware.py` when interface is firmed up.

### LINKED ISSUE #281: Implement Multi-Interface Initialisation and Management in CLI
Enhance the CLI and application to support the initialisation and management of multiple hardware interfaces within a workspace. This update will allow users to provide paths to existing hardware parameter files during workspace creation or be guided through a prompt to set up new interfaces if no paths are provided. Additionally, users will be able to add new interfaces to an existing workspace, either by providing a path to a hardware parameter file or through the interactive prompt.

### **Tasks**:

1. **Update CLI Workspace Initialisation**:
   - Modify the `initialise_app` method to accept paths to hardware parameter files when creating a new workspace.
   - If no paths are provided, prompt the user for the necessary information to create a new interface.
   - After the initial interface setup, prompt the user with the option: "Would you like to add another interface? (Y/N)". If the user selects "Y", repeat the interface setup process.
   - Store each interface’s configuration in a separate file named `hardware_params_<interface_name>.json` within the workspace directory.

2. **Add Interface to Existing Workspace**:
   - Implement a new CLI command (`do_add_interface`) that allows users to add a new interface to an existing workspace.
   - This command should accept a path to a hardware parameter file or initiate the interactive prompt for creating a new interface.
   - The new interface should be saved using the `hardware_params_<interface_name>.json` naming convention and added to the active interfaces in the workspace.

3. **Implement Interface Listing and Details Commands**:
   - **`do_list_loaded_interfaces`**: Add a command to list all currently loaded interfaces in the workspace, displaying their names, levels, and any other relevant information.
   - **`do_show_interface`**: Add a command to show detailed information about a specific interface, such as its capabilities, level, and current job count.

4. **Handle Interface Activation/Deactivation (Future Enhancement)**:
   - Consider future functionality for activating and deactivating interfaces, allowing users to turn interfaces on or off without removing their configurations. This should be tracked as a separate issue.

5. **Error Handling**:
   - Ensure that the system gracefully handles cases where no valid hardware parameter files are found or provided.
   - Implement appropriate error messages if an interface file fails to load or if the user provides invalid input during the interactive prompt.

6. **Documentation**:
   - Update the documentation to reflect the new initialisation process, including how to add interfaces during workspace creation and how to add interfaces to an existing workspace.
   - Provide examples of how to use the new CLI commands to manage interfaces.

### **Acceptance Criteria**:
- The application should successfully create a new workspace with multiple interfaces based on user input or provided parameter files.
- Users should be able to add new interfaces to an existing workspace through the CLI.
- The `do_list_loaded_interfaces` command should accurately display information about the interfaces.
- All new functionality should be intuitive, with clear prompts and error messages.

### PR DESCRIPTION
**Pull request by mbjohns**
### Updated PR Description:

This PR expands and enhances the CLI for managing multiple hardware interfaces within a workspace. While it retains the core functionality outlined in the issue and original PR description, it also introduces several new features and adjustments to address practical requirements identified during implementation.

### Key Changes

1. CLI Workspace Initialisation
- Planned:
    - The `initialise_app` method now allows paths to hardware parameter files during workspace creation.
    - Prompts guide the user through creating new interfaces if no paths are provided.
    - Interfaces are saved in `hardware_params_<interface_name>.json` files.
- Implemented:
    - A rich, interactive workflow for interface setup, with improved user prompts and validations.
    - Users can choose between an interactive setup mode or loading parameters from an existing JSON file.
    - Interfaces can be assigned levels during setup, stored alongside other configurations.

2. Adding Interfaces to Existing Workspaces
- Planned:
	- A new CLI command (`do_add_interface`) to add interfaces interactively or via file input.
- Implemented:
	- `do_add_interface` includes detailed error handling for JSON file validation and user inputs.
	- Added support for dynamically updating and reinitialising the job manager when a new interface is added.

3. Interface Listing and Details Commands
- Planned:
	- A `do_list_loaded_interfaces` command to display interface names and details.
- Implemented:
	- Renamed to `do_list_interfaces`, with extended functionality:
	- Lists interface names, levels, hosts, users, and active job counts.
	- Displays job counts dynamically fetched from the job manager.

4. Submit Command Enhancements
- Planned:
	- None explicitly stated in the issue.
- Implemented:
	- Added a `--level` flag to the submit command, allowing users to target specific interface levels.
	- Extended input validation and enhanced the submission pipeline to support levels.

5. Documentation Updates
- Planned:
	- Update documentation with new workflows and examples.
- Implemented:
	- Significantly expanded CLI walkthroughs with detailed examples for workspace initialisation, interface addition, job submission, and interface listing.
	- Added notes for advanced CLI usage, such as specifying interface levels.

### Divergences and Additional Features

- Enhanced Prompts: Improved user prompts and validations, ensuring intuitive and (hopefully) error-resilient workflows.
- Dynamic Job Manager Updates: The job manager is now re-initialised when interfaces are added, ensuring seamless operation.
- Extended Submit Command: The `--level` flag was not part of the original scope but enhances flexibility in targeting hardware interfaces.

### Remaining Tasks
- Consider if a `do_show_interface` command for detailed interface insights is necessary.
- Consider whether to add interface activation/deactivation functionality (deferred for future enhancement).

### Acceptance Criteria
- The application now supports creating workspaces with multiple interfaces via interactive or file-based inputs.
- Users can add interfaces to existing workspaces, specify levels, and list details through the CLI.
- Documentation has been updated to reflect changes, with examples.

*Note to reviewer: Sorry this is such a monster PR!*

Closes #281 & Closes #83

### CODE REVIEW

File: exauq/app/cli.py [hunk 1]
```diff
@@ -239,80 +354,344 @@ def initialise_app(self) -> None:
...
+        """Prompt the user to configure a hardware interface."""
+        self._generate_bordered_header(
+            "Hardware Interface Configuration details", border_char="-"
+        )
+        for param, prompt in factory.interactive_prompts.items():
+            value_str = input(f"{prompt}: ")
+            try:
+                factory.set_param_from_str(param, value_str)
+            except ValueError as e:
+                self._render_error(f"Invalid value -- {e}")
```
[the remark this answers is not in the document]
Comment (mbjohns) [reply exauq/app/cli.py]: Thank you for your detailed feedback! I've addressed the concerns by adding a validation method within the CLI class to ensure interface names are unique and follow valid naming rules. The interactive prompt now allows users to retry invalid inputs immediately, so they don't have to restart the whole process. The `_validate_interface_name` method catches duplicate names, ensuring a smoother experience and preventing the bug you highlighted.

File: exauq/app/startup.py [hunk 2]
```diff
@@ -454,3 +457,45 @@ def parse(x: str) -> Optional[bool]:
             return False
 
     return parse
+
+
+def make_int_parser(
```
[the remark this answers is not in the document]
Comment (mbjohns) [reply exauq/app/startup.py:420]: Good suggestion. Done.

**Review by HarryWhiteRSE — CHANGES_REQUESTED** [review]
Still a bit of an issue with the repeated name. It is fine upon initialisation of the workspace and it gracefully requests a new name. I've added a few suggestions as to potential bug fixes for implementing this with a file and also for the `add_interface` command. I would certainly recommend investigating the `factory` -> `factory_add` one further to really see what is going on there. 

Everything else looks good and the interactive prompt is much nicer to use alongside a much neater summary table! 🚀

File: exauq/app/cli.py [hunk 3]
```diff
@@ -239,80 +362,436 @@ def initialise_app(self) -> None:
...
+        for param, prompt in factory.interactive_prompts.items():
+            while True:
+                value_str = input(f"{prompt}: ")
+                try:
+                    if param == "name":
+                        self._validate_interface_name(value_str)
+                    factory.set_param_from_str(param, value_str)
+                    break
+                except ValueError as e:
+                    self._render_error(f"Invalid value -- {e}")
```
[the remark this answers is not in the document]
Comment (mbjohns) [reply exauq/app/cli.py:740]: Thanks for the suggestion! I agree that a positive integer check is necessary. I've implemented a `make_pos_int_parser` method in the `startup` module and applied it to the `level` parameter in `UnixServerScriptInterfaceFactory`. Ideally, I'd like to move the name validation logic to the factory class, but I need to find a clean solution for tracking existing interface names and ensuring the factory is aware of them. I'll revisit this once I've worked that out.

File: exauq/app/cli.py [hunk 4]
```diff
@@ -239,80 +362,436 @@ def initialise_app(self) -> None:
...
+        """Prompt the user to enter a positive integer, with error handling for invalid inputs."""
+        while True:
+            try:
+                value = int(input(prompt))
+                if value <= 0:
+                    raise ValueError("Value must be greater than zero.")
+                return value
+            except ValueError as e:
+                self._render_error(f"{error_message} ({e})")
+
```
[the remark this answers is not in the document]
Comment (mbjohns) [reply exauq/app/cli.py:551]: See above.

File: exauq/app/cli.py [hunk 5]
```diff
@@ -337,20 +832,167 @@ def _make_submissions_table(self, jobs: tuple[Job]) -> str:
...
+            display_name, factory_cls = self._select_hardware_interface_prompt()
+            factory = factory_cls()
+            factory.load_hardware_parameters(args.file.name)
+            self._finalise_hardware_setup(factory, factory_cls)
+
+        else:
+            display_name, factory_cls = self._select_hardware_interface_prompt()
+            factory = factory_cls()
+            self._hardware_interface_configuration_prompt(factory)
+            self._finalise_hardware_setup(factory, factory_cls)
```
[the remark this answers is not in the document]
Comment (mbjohns) [reply exauq/app/cli.py:862]: Thanks for the suggestion! I've implemented something similar, incorporating a few tweaks along the way.

**Review by HarryWhiteRSE — APPROVED** [review]
LGTM! 🚀 

This is a great addition to the job management side of the Toolbox! I think it would be worth opening a more long-term issue to refactor some of this due to the size of the CLI class, however, all the functionality is there and it now feels very robust!

**Pull request merged**